import.go

```diff
@@ -45,57 +82,121 @@ func newImporter(tree *MutableTree, version int64) (*Importer, error) {
 		return nil, errors.New("tree must be empty")
 	}
 
-	return &Importer{
-		tree:    tree,
-		version: version,
-		batch:   tree.ndb.db.NewBatch(),
-		stack:   make([]*Node, 0, 8),
-		nonces:  make([]uint32, version+1),
-	}, nil
+	importer := &Importer{
+		tree:       tree,
+		version:    version,
+		batch:      tree.ndb.db.NewBatch(),
+		stack:      make([]*Node, 0, 8),
+		nonces:     make([]uint32, version+1),
+		batchMtx:   sync.RWMutex{},
+		chanConfig: DefaultChannelConfig(),
+	}
+
+	importer.chanConfig.chNodeDataWg.Add(1)
+	go setBatchData(importer)
+
+	importer.chanConfig.chBatchWg.Add(1)
+	go batchWrite(importer)
+
+	return importer, nil
+}
+
+// setBatchData get the next serialized node data from channel, and write the data to the current batch
+func setBatchData(i *Importer) {
+	for i.batch != nil {
+		if nodeData, open := <-i.chanConfig.chNodeData; open {
+			i.batchMtx.RLock()
+			if i.batch != nil {
+				err := i.batch.Set(i.tree.ndb.nodeKey(nodeData.node.GetKey()), nodeData.data)
+				if err != nil {
+					i.batchMtx.RUnlock()
+					i.chanConfig.chError <- err
+					break
+				}
+			}
+			i.batchSize++
+			i.batchMtx.RUnlock()
+			// Only commit a new batch if size meet desiredBatchSize and there's no pending batch write or we exceed maxBatchSize
+			if (i.batchSize >= i.chanConfig.desiredBatchSize && len(i.chanConfig.chBatch) < 1) || i.batchSize >= i.chanConfig.maxBatchSize {
+				i.chanConfig.chBatch <- i.batch
+				i.batch = i.tree.ndb.db.NewBatch()
+				i.batchSize = 0
+			}
+		} else {
+			break
+		}
+	}
+	i.chanConfig.chNodeDataWg.Done()
+}
+
+// batchWrite get a new batch from the channel and execute the batch write to the underlying DB.
+func batchWrite(i *Importer) {
+	for i.batch != nil {
+		if nextBatch, open := <-i.chanConfig.chBatch; open {
+			err := nextBatch.Write()
+			if err != nil {
+				i.chanConfig.chError <- err
+				break
+			}
+			i.batchMtx.Lock()
+			nextBatch.Close()
+			i.batchMtx.Unlock()
+		} else {
+			break
+		}
+	}
+	i.chanConfig.chBatchWg.Done()
 }
 
 // writeNode writes the node content to the storage.
 func (i *Importer) writeNode(node *Node) error {
+
+	if err := i.validate(node); err != nil {
+		return err
+	}
+
+	if nodeData, err := i.getDataBytes(node); err != nil {
+		return err
+	} else {
+		// Handle the remaining steps in a separate goroutine
+		i.chanConfig.chNodeData <- NodeData{
+			node: node,
+			data: nodeData,
+		}
+	}
+
+	return nil
+}
+
+func (i *Importer) validate(node *Node) error {
 	if _, err := node._hash(node.nodeKey.version); err != nil {
 		return err
 	}
 	if err := node.validate(); err != nil {
```

Comment:
Or you can simply invoke: return node.validate()
Then delete the following 4 lines